### PR TITLE
[babel-preset-expo] Added support for jsxRuntime option

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

Now that we support React 17, we can add support for their big feature: [auto JSX transforms](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html).

# How

The React Native preset does not expose this functionality, so I opted to disable the transforms entirely, and add them back with customizations.

I documented my findings in-depth since Babel can be a bit confusing.

### Eslint

To disable eslint warnings, create a `.eslintrc.js` and add:

```js
module.exports = {
  rules: {
    // Disable React JSX transform warnings because
    // `jsxRuntime: 'automatic'` is passed to `babel-preset-expo`.
    'react/jsx-uses-react': 'off',
    'react/react-in-jsx-scope': 'off',
  },
};
```

### TypeScript

> Requires TypeScript 4.1

In your `tsconfig.json`, add the [new jsx setting](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#jsx-factories):

```json
{
  "compilerOptions": {
    "jsx": "react-jsx"
  },
  "extends": "expo/tsconfig.base"
}
```

### Flow

[idk](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#flow).

# Test Plan

- Wrote unit tests, but pending https://github.com/expo/expo/pull/13946

- Run `expo start -c` in bare-sandbox, it should load up as expected.
- Disable the babel `jsxRuntime` option, and try again, it should produce an error "Can't find variable: React", this is because I removed React from the App.js component to test it.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by ide — COMMENTED** [review]
The feature/code looks OK but we should name the property something more context-insensitive like jsxRuntime. Otherwise if you told me the Babel preset had an option called runtime, I'd guess the values would be something like "expo" vs "react-native" vs "ionic".

**Review by ide — COMMENTED** [review]
Unrelated to this PR but heads up that the default Metro Babel preset that comes with RN 0.64 breaks part of the JS spec here: https://github.com/facebook/metro/blob/2a0f77925e3299e01c2bfe82d0356145a643adc0/packages/metro-react-native-babel-preset/src/configs/main.js#L107. We'll need to remove `{ loose: true }`.

**Review by Kudo — COMMENTED** [review]
dunno the exact reason, but looks like metro had changed `Jsx` to `JSX` in [this commit](https://github.com/facebook/metro/commit/c18088e276dc1a66e44b248b955fb12cc750169f) and react-native 0.64 depends on this new version of metro-react-native-babel-preset 🤔

File: packages/babel-preset-expo/__tests__/index-test.js [hunk 1]
```diff
@@ -44,6 +45,30 @@ import { View } from 'react-native';
     expect(code).toMatchSnapshot();
   });
 
+  it(`supports automatic runtime`, () => {
```
Comment (ide) [inline packages/babel-preset-expo/__tests__/index-test.js:48]:
```suggestion
  it(`supports automatic JSX runtime`, () => {
```

File: packages/babel-preset-expo/index.js [hunk 2]
```diff
@@ -22,6 +24,34 @@ module.exports = function(api, options = {}) {
...
+  // we override this logic outside of the metro preset so we can add support for
+  // React 17 automatic JSX transformations.
+  // If the logic for `useTransformReactJSXExperimental` ever changes in `metro-react-native-babel-preset`
+  // then this block should be updated to reflect those changes.
+  if (!platformOptions.useTransformReactJSXExperimental) {
+    extraPlugins.push([
+      require('@babel/plugin-transform-react-jsx'),
+      {
+        // Defaults to `classic`, pass in `automatic` for auto JSX transformations.
+        runtime: options && options.jsxRuntime,
```
Comment (ide) [inline packages/babel-preset-expo/index.js:47]:
```suggestion
        runtime: options?.jsxRuntime,
```

**Review by ide — APPROVED** [review]

File: packages/babel-preset-expo/index.js (hunk 2 above)
Comment (EvanBacon) [reply packages/babel-preset-expo/index.js:47]: It's JS

**Review by brentvatne — APPROVED** [review]

**Review by Kudo — APPROVED** [review]
showing the backward compatible warning is neat 👍

**Pull request merged**